### ISSUE TITLE
Support if..else syntax?

### ISSUE THREAD

**Issue by vutrungduc7593** [scijava/parsington#6]
Hi, i want to know whether the lib support if..else operator or does it can simulate it? Thanks.

**Comment by ctrueden** [comment]
Parsington does not currently support [ternary operators](https://en.wikipedia.org/wiki/%3F:), only unary and binary. You could [extend the shunting yard implementation to add it](https://stackoverflow.com/q/35609168/1207769). A pull request doing this would be warmly welcome, and I would make it a priority to get it reviewed and merged.

**Comment by CodeMaven** [comment]
Does this actually require true ternary operators? I was thinking about supporting "condition?then:else" constructs and it seems to me that this could be done by implementing a conditional operator '?' and a branch operator ':' . When evaluating the ? wouldn't really do anything but eval the left side first and put a boolean onto the stack then the : operator on the right side of the ? tree would then eval only one of the branches and throw away the other. Semantically speaking, the : is only valid following a ?. I think I'm going to need this functionality for a project I'm working on so I might take a stab at it.

**Comment by CodeMaven** [comment]
Then again.... There is currently no deferred parsing in Parsington so doing it this way could never support short circuiting - both branches would always be parsed which makes it only useful for the comm assignment case of:
a = condition?valA:valB
which I would think is the main use case, but there may be situations where you want to do something else completely.... like
condition?b=a:c=a
If you have assignments inside the branches then without any way to do deferred parsing they would both always be evaluated. I can't think of a _clean_ way to support differed parsing off of the top of my head, though I can think of possibly an untidy way of doing it with some ugly logic in the evaluator that would feel very hacky to me...

**Comment by ctrueden** [comment]
@CodeMaven Thanks for your thoughts, and sorry for not responding earlier.

Firstly: did you find a way to implement what you need for your project? I'm always curious what other solutions are out there—especially if there is something else better that makes a project I maintain obsolete. :wink:

As for the deferred parsing: I do not think deferred _parsing_ is necessary—only deferred _evaluation_ if you want to achieve short-circuiting behavior. The easiest way I could see to achieve deferred evaluation was via a new `Evaluator` implementation that operates on syntax trees (recursively) rather than postfix queues (iteratively).

I have pushed a branch called `ternary` with working code. Check out the `DefaultTreeEvaluator`, or run the `Main` which now uses it by default.

I have not merged this to master because: A) it needs testing; B) there are no unit tests. I do not have time to polish this up any further at the moment, but feedback from the community here would be most welcome.

**Comment by CodeMaven** [comment]
Hi @ctrueden,

I have not worked on this yet. It is on my backlog, quite deep down. The project that I am working on is massive so I've put this aside while I work on other aspects. I am not using the parsington evaluator, just the parser.   My project is a financial trading system and the parser is being used to evaluate user created expressions.   However, instead of an evaluator I'm actually passing the output from parsington into a code generator that I've written which is generating java bytecode on the fly to implement the expression as a functional predicate. The generated code is used in a stream for filtering large result sets. Specifically I do need this feature for certain use-cases where my filtering logic has to look at different fields depending on other fields. I.e:

`interestType=="FLOATING"?floatingRate>0:fixedRate>0
`  

I actually would like to release my little predicate generator / compiler a opensource, but again I just haven't gotten around to it. There are a couple of missing features and it's somewhat tightly coupled with the rest of my project at the moment because it generates calls to retrieve values directly from my cache.

Cheers,
Troy

**Comment by ctrueden** [comment]
> I'm actually passing the output from parsington into a code generator that I've written which is generating java bytecode on the fly to implement the expression as a functional predicate.

In that case, I think you can achieve what you need simply by including `?` and `:` binary operators at precedence 1 and 1.5, respectively, as I've done on the `ternary` branch. Does your code generator consume postfix queues, or syntax trees? In the syntax tree, for nodes with the operator `?`, the left child should be a boolean expression, and the right child should be the `:` operator, whose left and right children are the true and false consequent expressions, respectively. Hopefully the mapping from that structure to bytecode is straightforward?

**Comment by CodeMaven** [comment]
At the moment it's using syntax trees, but I was hoping to move to a postfix queue. Since the JVM itself is a stack machine the postfix queue (in theory) is the most straight-forward way to do it. I had started with using a postfix queue but had to switch to using the syntax tree to solve an issue that I had early on.. I don't remember exactly what my issue was, but I think it was to do with type casting - I needed to look down both branches of the tree to see if a cast would be required. However, I'm pretty sure that with a bit of refactoring I can do it with the postfix queue. logically, that makes the code generation very straightforward in most cases since simple asthmatic is perfectly 1 to 1 in bytecode with the postfix queue,  i.e. 
```
expression: "a = 3 + 5"
parse queue: a 3 5 + =
Generated bytecode (assembly):
1 (variable index)
iconst_3
iconst_5
iadd
istore


```

**Comment by ctrueden** [comment]
@CodeMaven One option would be to start with a syntax tree, do some preprocessing to smooth out any rough edges, and then dump the tree as a postfix queue. This would be an easy way e.g. to transform the `?:` two binary operators into a single true ternary operator.

**Comment by talwgx** [comment]
I've recently begun using parsington for a project I'm working on, and it's really great! To the topic of ternary operations: I would actually benefit from that option and would love to see it incorporated into the main branch.

**Comment by ctrueden** [comment]
@talwgx I'm planning to work on this issue plus #10 and #16/#17 again in the coming days; will try to get a new release out soon.

**Comment by ctrueden** [comment]
@talwgx I merged the ternary work, and released parsington 2.1.0. It will be the last release in the 2.x series; I am now working on migrating the codebase to Java 8, and addressing the other open issues.
